drivers/gpio/gpio_pcf8574.c

```diff
@@ -0,0 +1,375 @@
+ * @brief clear some bits
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pins Pins which will be cleared
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error.
+ */
+static int pcf8574_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return pcf8574_port_set_raw(dev, (uint8_t)pins, 0, 0);
+}
+
+/**
+ * @brief Toggle some bits
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param pins Pins which will be toggled
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error.
+ */
+static int pcf8574_port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	return pcf8574_port_set_raw(dev, 0, 0, (uint8_t)pins);
+}
+
+/** Set the configured interrupts pin(s)*/
+static int pcf8574_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					   enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	const struct pcf8574_drv_cfg *drv_cfg = dev->config;
+	struct pcf8574_drv_data *drv_data = dev->data;
+
+	int rc;
+
+	if (!drv_cfg->gpio_int.port) {
+		return -ENOTSUP;
+	}
+
+	/* This device supports only edge-triggered interrupts. */
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		return -ENOTSUP;
+	}
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	return rc;
+}
+
+/** Register the callback in the callback list*/
+static int pcf8574_manage_callback(const struct device *dev, struct gpio_callback *callback,
+				   bool set)
+{
+	struct pcf8574_drv_data *drv_data = dev->data;
+
+	return gpio_manage_callback(&drv_data->callbacks, callback, set);
+}
+
+/** Initialize the pcf8574*/
+static int pcf8574_init(const struct device *dev)
+{
+	const struct pcf8574_drv_cfg *drv_cfg = dev->config;
+	struct pcf8574_drv_data *drv_data = dev->data;
+	int rc;
+
+	if (!device_is_ready(drv_cfg->i2c.bus)) {
+		LOG_ERR("%s is not ready", drv_cfg->i2c.bus->name);
+		return -ENODEV;
+	}
+
+	/* If the INT line is available, configure the callback for it. */
+	if (drv_cfg->gpio_int.port) {
+		if (!device_is_ready(drv_cfg->gpio_int.port)) {
+			LOG_ERR("Port is not ready");
+			return -ENODEV;
+		}
+
+		rc = gpio_pin_configure(drv_cfg->gpio_int.port, drv_cfg->gpio_int.pin,
+					drv_cfg->gpio_int.dt_flags | GPIO_INPUT);
+		if (rc != 0) {
+			LOG_ERR("%s: failed to configure INT line: %d", dev->name, rc);
+			return -EIO;
+		}
+
+		rc = gpio_pin_interrupt_configure(drv_cfg->gpio_int.port, drv_cfg->gpio_int.pin,
+						  GPIO_INT_EDGE_TO_ACTIVE);
+		if (rc != 0) {
+			LOG_ERR("%s: failed to configure INT interrupt: %d", dev->name, rc);
+			return -EIO;
+		}
+
+		gpio_init_callback(&drv_data->int_gpio_cb, pcf8574_int_gpio_handler,
+				   BIT(drv_cfg->gpio_int.pin));
+		rc = gpio_add_callback(drv_cfg->gpio_int.port, &drv_data->int_gpio_cb);
+		if (rc != 0) {
+			LOG_ERR("%s: failed to add INT callback: %d", dev->name, rc);
+			return -EIO;
+		}
+	}
+
+	/* Device configured, unlock it so that it can be used. */
+	k_sem_give(&drv_data->lock);
```

Comment:
The `drv_data->lock` isn't initialized with `k_sem_init`. Also, if something went wrong in this function, the lock could never be taken and would block any call from the application.